adaptlearning/adapt_framework
# issue/2747 Added questionModel attemptStates API

#2747 

To keep the previous question attempt states for use in plugins.

Requires [#192](https://github.com/adaptlearning/adapt-contrib-spoor/pull/192) spoor pr to be saved and restored across sessions.

#### Added
* `trackableType` property throughout to list the type of the `trackable` attributes
* `questionModel.getAttemptState` Returns the raw state of the component in a format for spoor to store
* `questionModel.getAttemptObject` Returns a formatted state object which can be passed through to `questionModel.set`
* `questionModel.setAttemptObject` Sets the current state of the component from a state object
* `questionModel.addAttemptObject` Adds an attempt object to the attemptStates store
* `questionModel.getAttemptObjects` Returns an array of the component attempt state objects
* `_spoor._tracking._shouldStoreAttempts` to `config.json`
* Question completion automatically adds a state object to the attempt state store

#### Testing
* Add component 'c-60' mcq `_isResetOnRevisit: "soft"` 
* Keep attempting 'c-60'
* Use the console to determine mcq attempts
```js
var Adapt = require('core/js/adapt');
Adapt.findById('c-60').getAttemptObjects();
```
* To see states restored across sessions use spoor pr [#192](https://github.com/adaptlearning/adapt-contrib-spoor/pull/192)
* And enable `_spoor._tracking._shouldStoreResponses` and `_spoor._tracking._shouldStoreAttempts` in `config.json`
